### PR TITLE
DMatrix optimizations

### PR DESCRIPTION
**Pull request by ShvetsKS**
Observed https://github.com/dmlc/xgboost/issues/5530 slowdown in DMatrix creation was fixed.

Performance improvements:
Data set | DMatrix create time(master), s | DMatrix create time | Speedup
-- | -- | -- | --
airline-ohe with pandas DF | 13.25 |  8.3 |  1.6
airline-ohe as contiguous numpy| 6.6 |  1.75 |  3.7
higgs1m as contiguous numpy| 0.28 |  0.077 |  3.6

"Dmatrix create + Prediction" improvements (**depth = 1, n_iterations = 10**):
  | airline-ohe with pandas DF | airline-ohe contiguous numpy | higgs1m contiguous numpy
-- | -- | -- | --
Master | 13.37 | 6.8 | 0.296
Dmatrix Opt | 8.44 | 1.86 | 0.085
Speedup | 1.58 | 3.65 | 3.4

### CODE REVIEW

File: src/data/data.cc [hunk 1]
```diff
@@ -865,6 +866,87 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
...
+  const size_t thread_size = (expected_rows+1)/nthread;
+  const size_t tail_size = (expected_rows+1) - thread_size*nthread;
+
+  builder.InitBudget(thread_size, nthread, tail_size);
+  uint64_t max_columns = 0;
+
+  // First-pass over the batch counting valid elements
+  size_t num_lines = batch.Size();
+  if (num_lines == 0) {
+    return max_columns;
```
Comment (SmirnovEgorRu) [inline src/data/data.cc:902]: You don't call ```omp_set_num_threads(nthread_original);``` in this case

File: src/data/data.cc [hunk 2]
```diff
@@ -865,6 +866,87 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
   return max_columns;
 }
 
+template<>
+uint64_t SparsePage::Push<xgboost::data::DenseAdapterBatch>(
+               const xgboost::data::DenseAdapterBatch& batch,
+               float missing, int nthread, int nthread_original) {
```
Comment (SmirnovEgorRu) [inline src/data/data.cc:872]: We can make the changes in general without templates specification.

Comment (ShvetsKS) [reply src/data/data.cc:872]: unfortunately there is no possibility to keep obtained speed up for unified version.
The problem is: 
`builder.InitBudget(expected_rows, nthread);` allocates `nthread` times bigger buffer (on datasets with small number of columns like higgs it becomes the hot spots and no performance improvements are observed).

To eliminate described problem in specification optimization was possible by taking into account batch adapter type (threading by rows was implemented), it won't be applicable for `CSCAdapterBatch`.

Comment (ShvetsKS) [reply src/data/data.cc:872]: But general case will be also optimized :)

File: src/data/data.cc (hunk 1 above)
Comment (ShvetsKS) [reply src/data/data.cc:906]: fixed

**Comment by ShvetsKS** [comment]
Hi @trivialfis, is it possible to change the default value of `nthread` parameter in DMAtrix initialization?https://github.com/dmlc/xgboost/pull/5877/files#diff-fd53d68e0037d3512896122d1248d969R448

as I understand default parameter `nthreads` for DMAtrix creation was set to  `1` (https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/core.py#L425) as there was a negative scaling with increasing number of threads in `uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread)` method? 
It was changed to `-1` as currently we can see speed up with  threading.

Thanks you in advance.

fix of problem with omp max reduction on VS is in progress.

File: src/data/data.cc (hunk 2 above)
Comment (SmirnovEgorRu) [reply src/data/data.cc]: If you need to change only one call depending on template type - you can use small separate template function which will call InitBudget in appropriate manner based on Adapter type.

Comment (ShvetsKS) [reply src/data/data.cc]: unified approach was implemented.

**Comment by ShvetsKS** [comment]
Not sure that failed step "continuous-integration/travis-ci/pr" is related to current changes.
@trivialfis , @hcho3  sorry for bothering you, should I rerun it?

File: src/common/group_data.h [hunk 3]
```diff
@@ -74,6 +84,34 @@ class ParallelGroupBuilder {
     }
     trptr[offset_key] += nelem;
   }
+
+
+
+  /*! \brief step 3: initialize the necessary storage */
```
Comment (trivialfis) [inline src/common/group_data.h:90]: So we have 2 different step 3 now?

Comment (ShvetsKS) [reply src/common/group_data.h:90]: @trivialfis thanks for review

So, for Airline-ohe data the majority of speed up comes from increased number of threads (with applied fix to remove false sharing). No significant improvements from specialization (~10%):

 DMatrix create time,s | Master  | This PR | This PR without specialized (`InitStorage`,`InitBudged`)
-- | -- | -- | --
airline-ohe contiguous numpy | 6.7 | **1.8** | **1.99**

But for Higgs data increasing number of threads (without specialized `InitStorage`,`InitBudged` method) leads to unnecessary increasing of allocated buffers sizes and time to fill and work with it.
As a result only specialization could help in this case to get declared improvements:
 DMatrix create time,s | Master  | This PR | This PR without specialized (`InitStorage`,`InitBudged`)
-- | -- | -- | --
higgs1m contiguous numpy | 0.28 | **0.08** | **0.24**

Without `InitStorage`,`InitBudged` specialization it's impossible to apply more effective memory usage as threading is implemented for abstract lines(columns or rows).

Different implementation of `InitStorage` method is required to handle optimized case (common buffer was allocated for all threads).

Comment (ShvetsKS) [reply src/common/group_data.h:90]: @trivialfis as we have the case where specialization gives significant improvement it would be better to keep it.

Comment (trivialfis) [reply src/common/group_data.h:90]: Wow, that's a factor of 4.  However, I would still want to minimize the specialization (maybe I can help).  I have been going through the code base and trying to unify the code.  A few examples:

- We have 4 different enumeration code doing similar things.  This is a huge blocker for categorical data.
- We have 4 different partitioning code.  This is a huge blocker for adaptive variate tree.
- We have 6 entries for quantile sketching.  Blocker for categorical data.
- We have 4 different histogram building code, blocker for multi target tree.
- We have 2 column sampler, blocker for feature weights.

I can go on with other features that we can have but haven't been able to do.  But as you can see if we keep specializing code XGBoost will eventually sink because failing to catch up with the machine learning algorithms (and XGBoost is all about machine learning algorithm).  That's the primary reason I haven't been able to allocate time to optimize our GPU distributed pipeline.  Because the major target is machine learning.  If someone have a better learning algorithm, or simply better heuristic, all the optimization doesn't matter because the new algorithm will beat you, 2 more percent of accuracy can overweight all the optimization (people go with pure Python library instead of optimized libraries for features, they can simply use a cloud to scale the training when needed with a bit more money).  And it would be difficult for us to catch up with the algorithm development.

For the case of DMatrix, this seems irrelevant.  But I have a strong feeling that it might eventually lead to issues.  For example, there are algorithms for determining data encoding by cross validation at early pipeline, DMatrix might be the place for it to happen (I hope it will be sketching, but I don't know yet).  So please keep in mind that, every specialization has its cost.  I made the mistake for CPU sketching, now I have to spent my time to fix my mess and your time to review it.

Comment (trivialfis) [reply src/common/group_data.h:90]: Not trying to be a pain here.  The idea I want to convey is, optimization is absolutely awesome, but we should make sure that there's no other way than specialization.  For example, in general pipeline, float64 is the default precision for numpy.  `ctypes` has to allocate an array of `float32` and iterate through it to make a copy before passing it to XGBoost, which is quite expensive. (I vaguely remember for a large scipy csr_matrix it takes half of the initialization time, and double the memory usage).  Making a API accepting a general void data buffer can remove that in the pipeline.  Also, for `hist` algorithm and possibly `approx`, constructing DMatrix can be completely skipped because it uses the gradient index for training instead of accessing data itself.  So I believe there are plenty of ways to make optimization without making these small specializations.  When specialization is indeed needed, we can limit the scale.  For example the `SetIndexDataForSparse` function currently in XGBoost can be completely merged with dense version if it accept an functor for getting index with offset, the functor would be a one liner.

Comment (ShvetsKS) [reply src/common/group_data.h]: @trivialfis I totally agree with you. So I reverted specification of `InitBudged/Storage` methods to simplify support of the code.
So we have next improvements results:

DMatrix create time,s | Master | Optimized | Speed up | Optimized without specification | Speed up
-- | -- | -- | -- | -- | -- 
airline-ohe contiguous numpy | 6.6 | 1.8 | 3.67 |1.99 | 3.32
msrank10k | 3.9 | 1.16 | 3.36 | 1.61 | 2.42
mortgage1Q | 4.06 | 1.1 | 3.69 | 2.93 | 1.39

Comment (ShvetsKS) [reply src/common/group_data.h]: @trivialfis It's a great idea to skip DMatrix construction for `hist` method,  I think we could start to implement it.
And can we consider this option for prediction stage also? In some cases DMatrix creation take more than half of inference time (it's very important to have fast prediction in deployed application). Would it be applicable if we introduce possibility to perform inference from pandas/numpy directly without any memcopy?

Otherwise we have potential benefits from specialization and I'd propose to have `ParallelGroupBuilder` specialization for adapters, or methods provided by `ParallelGroupBuilder` class should be as part of concrete `AdapterBatch` class for example.

**Comment by ShvetsKS** [comment]
@trivialfis could you take a look at failed tests, I'm not sure that it's related to this PR.
there is a problem with `DeviceQuantileDMatrix `, and fails: `Multiple processes within communication group running on same CUDA device is not supported`. After I look at the code, seems current changes should not affect gpu implementation.

and I can hardly believe that 'Error in loadNamespace(name) : there is no package called 'DiagrammeR' ' in r test is related to current changes.

File: src/common/group_data.h (hunk 3 above)
Comment (trivialfis) [reply src/common/group_data.h]: > And can we consider this option for prediction stage also?

No, you can run prediction on training dataset with gradient index, but not any other dataset.  But we do have an inplace prediction function.

**Comment by trivialfis** [comment]
Let me take a look later.  This is new.

**Comment by ShvetsKS** [comment]
> Let me take a look later. This is new.

in another PR(temporary) with the same changes: https://github.com/dmlc/xgboost/pull/6007 step with r tests (test-r-3.5.3) have been passed.

**Comment by ShvetsKS** [comment]
@trivialfis could you approve this PR, as seems there is no real problem with testing?

**Comment by trivialfis** [comment]
I'm debugging the test failure on jenkins.  As it's blocking everything so I will come back to other issues later.

File: src/data/data.cc [hunk 4]
```diff
@@ -829,48 +830,68 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
...
+    return max_columns;
+  }
+  std::vector<std::vector<uint64_t>> max_columns_vector(nthread);
 
   // First-pass over the batch counting valid elements
-  size_t batch_size = batch.Size();
-#pragma omp parallel for schedule(static)
-  for (omp_ulong i = 0; i < static_cast<omp_ulong>(batch_size);
-       ++i) {  // NOLINT(*)
+#pragma omp parallel num_threads(nthread)
```
Comment (trivialfis) [inline src/data/data.cc:844]: Please help adding `dmlc::OMPException` to capture exception happened inside omp block.

File: src/data/data.cc [hunk 5]
```diff
@@ -829,48 +830,68 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
...
+
   builder.InitStorage();
 
   // Second pass over batch, placing elements in correct position
-#pragma omp parallel for schedule(static)
-  for (omp_ulong i = 0; i < static_cast<omp_ulong>(batch_size);
-       ++i) {  // NOLINT(*)
+
+#pragma omp parallel num_threads(nthread)
+  {
```
Comment (trivialfis) [inline src/data/data.cc:878]: Please help adding `dmlc::OMPException` to capture exception happened inside omp block.

File: src/data/data.cc [hunk 6]
```diff
@@ -829,48 +830,68 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
           last_line.GetElement(last_line.Size() - 1).row_idx - base_rowid;
     }
   }
-  builder.InitBudget(expected_rows, nthread);
+  size_t batch_size = batch.Size();
+  const size_t thread_size = batch_size/nthread;
```
Comment (trivialfis) [inline src/data/data.cc:834]: Do you need to roundup?

Comment (ShvetsKS) [reply src/data/data.cc:834]: with current `begin`, `end` calculation there is no need to roundup

File: src/data/data.cc (hunk 5 above)
Comment (ShvetsKS) [reply src/data/data.cc:878]: applied

File: src/data/data.cc (hunk 4 above)
Comment (ShvetsKS) [reply src/data/data.cc:844]: fixed

**Comment by ShvetsKS** [comment]
> LGTM overall. Could you please help addressing some minor issues?

@trivialfis  thanks for review

**Comment by ShvetsKS** [comment]
@trivialfis is failed `codecov/patch` check critical for merging?

**Review by trivialfis — APPROVED** [review]

**Comment by trivialfis** [comment]
@ShvetsKS Thanks for the good work and patience!
